pyccel/parser/extend_tree.py

```diff
@@ -101,15 +101,12 @@ def get_last_lineno(ast):
             ast = ast.body[-1]
     return ast.lineno
 
-def insert_comments(ast, comment_lines_no, comments, else_no, attr='body', col_offset = None):
+def insert_comments(ast, comment_lines_no, comments, else_no, attr='body'):
     if len(comments) == 0:
         return
     body        = getattr(ast, attr)
 
-    if attr == 'orelse':
-        assert col_offset is not None
-    else:
-        col_offset = body[0].col_offset
+    col_offset = body[0].col_offset
```

Comment:
the col_offset is not set properly sometimes by the Python parser and different Python versions give different results,  you think that we don't need that?

Reply:
> the col_offset is not set properly sometimes by the Python parser

Interesting, I did not know that. I know that the `col_offset` is not set for ast objects such as `Module`, but then they are not needed here.

> different Python versions give different results

As long as the numbers are consistent within a given version, I don't see this as a problem

> you think that we don't need that?

The reason I changed this is because the following code:
```
def f():
    # 1
    if True:
        # 2
        pass
    else:
        # 3
        for i in range(4):
            # 4
            pass
        # 5
    # 6
```
visits `insert_comments` and sets `col_offset` to the following values:
- `insert_comments(Module, ..., attr='body',...)` sets `col_offset=0` (ok)
- `insert_comments(FunctionDef, ..., attr='body',...)` sets `col_offset=4` (ok)
- `insert_comments(If, ..., attr='body',...)` sets `col_offset=8` (ok)
- `insert_comments(If, ..., attr='orelse',...)` sets `col_offset=4` (wrong)
- `insert_comments(For, ..., attr='body',...)` sets `col_offset=12` (ok)

so the only time we set `col_offset` to something other than `body[0].col_offset` we are setting it to the wrong value

Reply:
The alternative fix would be to modify [this line](https://github.com/pyccel/pyccel/blob/master/pyccel/parser/extend_tree.py#L192) and [this line](https://github.com/pyccel/pyccel/blob/master/pyccel/parser/extend_tree.py#L248) to pass the correct `col_offset`. But I don't know how to calculate that without checking the `col_offset` of items in the body. In the example above the additional each line increases the offset by 4, but that is due to my preferred indenting not anything inherent to python

Reply:
yes,  maybe it's better to add the [unparseast](https://github.com/simonpercivall/astunparse) module to generate the string back to be able to compare it with the original.
because I tried something like
```python
if True:   
    pass  
elif False: 
    pass  
else:  
    # 3  
    pass 
    # 4  
```
and it didn't work

Reply:
Interesting, this works for me with python 3.8

Reply:
I have Python 3.6.9, it puts the last comment after the else block instead of the end of the else block.

Reply:
```python

Reply:
we can do this
```python
class Unparser(astunparse.Unparser): 
    """Unparser the AST""" 
 
    def _CommentLine(self, node): 
        self.write('\n' + ' '*node.col_offset + node.s) 

```
and check that `Uparser(extend_tree(code)) == code`

Reply:
done